src/util.ts

```diff
@@ -1,82 +1,70 @@
-var deepEqual = require('fast-deep-equal');
-var isServer = process.title !== 'browser';
-exports.isServer = isServer;
-
-exports.asyncGroup = asyncGroup;
-exports.castSegments = castSegments;
-exports.contains = contains;
-exports.copy = copy;
-exports.copyObject = copyObject;
-exports.deepCopy = deepCopy;
-exports.deepEqual = deepEqual;
-exports.equal = equal;
-exports.equalsNaN = equalsNaN;
-exports.isArrayIndex = isArrayIndex;
-exports.lookup = lookup;
-exports.mayImpact = mayImpact;
-exports.mayImpactAny = mayImpactAny;
-exports.mergeInto = mergeInto;
-exports.promisify = promisify;
-exports.serverRequire = serverRequire;
-exports.serverUse = serverUse;
-exports.use = use;
-
-function asyncGroup(cb) {
+export const deepEqual = require('fast-deep-equal');
+export const isServer = process.title !== 'browser';
+
+export function asyncGroup(cb) {
   var group = new AsyncGroup(cb);
   return function asyncGroupAdd() {
     return group.add();
   };
 }
 
-/**
- * @constructor
- * @param {Function} cb(err)
- */
-function AsyncGroup(cb) {
-  this.cb = cb;
-  this.isDone = false;
-  this.count = 0;
-}
-AsyncGroup.prototype.add = function() {
-  this.count++;
-  var self = this;
-  return function(err) {
-    self.count--;
-    if (self.isDone) return;
-    if (err) {
+type ErrorCallback = (err?: Error) => void;
+
+class AsyncGroup {
+  cb: ErrorCallback;
+  isDone: boolean;
+  count: number;
+
+  constructor(cb: ErrorCallback) {
+    this.cb = cb;
+    this.isDone = false;
+    this.count = 0;
+  }
+
+  add() {
+    this.count++;
+    const self = this;
+    return function(err) {
+      self.count--;
+      if (self.isDone) return;
+      if (err) {
+        self.isDone = true;
+        self.cb(err);
+        return;
+      }
+      if (self.count > 0) return;
       self.isDone = true;
-      self.cb(err);
-      return;
-    }
-    if (self.count > 0) return;
-    self.isDone = true;
-    self.cb();
-  };
-};
+      self.cb();
+    };
+  }
+}
 
-/**
- * @param {Array<string | number>} segments
- * @return {Array<string | number>}
- */
-function castSegments(segments) {
-  // Cast number path segments from strings to numbers
-  for (var i = segments.length; i--;) {
-    var segment = segments[i];
-    if (typeof segment === 'string' && isArrayIndex(segment)) {
-      segments[i] = +segment;
-    }
+function castSegment(segment: string): string | number {
+  return (typeof segment === 'string' && isArrayIndex(segment))
+    ? +segment // sneaky op to convert numeric string to number
+    : segment;
+}
+
+export function castSegments(segments: Readonly<string | string[]>) {
+  if (typeof segments === 'string') {
+    return castSegment(segments);
   }
-  return segments;
+  // Cast number path segments from strings to numbers
+  return segments.map(segment =>
+    Array.isArray(segment)
+      ? castSegments(segment)
+      : castSegment(segment)
+  );
```

Comment:
Extra isArray check inside `map` and recursive call seems unnecessary, as long as `castSegments` is only ever called with a string or array-of-strings.
```suggestion
  return segments.map(segment => castSegment(segment));
```